_drafts/floating-point-101.md

```diff
@@ -0,0 +1,441 @@
+[a common effort](https://www.researchgate.net/publication/2954891_IEEE_754_An_Interview_with_William_Kahan) at the
+Institute of Electrical and Electronics Engineers (IEEE) to standardize floating-point arithmetic – the IEEE 754 working
+group. Two competing drafts prevailed: the Intel 8087 spec vs. the DEC VAX one. After some more arguments and error
+analysis, in 1981 Intel's draft won out, rapidly got adopted by everyone,
+and [the rest is history](https://www.intel.com/content/dam/www/public/us/en/documents/case-studies/floating-point-case-study.pdf)
+(though it took the committee another four years of bickering to publish that draft, of course).
+
+> For a detailed look at historic floating-point formats, see
+> [this great article by John Savard](http://www.quadibloc.com/comp/cp0201.htm).
+
+## The Specs
+
+Let's get into the meat of it, starting with… value sizes. This aspect informs the practical details of the scheme
+(range and precision), but _can't_ really be worked out from first principles. It's really just all about machine word
+sizes.
+
+Two floating-point formats are in common use today, and they both use the binary numeral system:
+
+-   32 bits, technically named `binary32`, but commonly **single precision**. Values of this size are called **floats**.
+-   64 bits, technically named `binary64`, but commonly **double precision**. Values of this size are called
+    **doubles**.
+
+Back in the day, double precision was reserved for operations requiring exceptional precision or range. With 64-bit
+processor architectures becoming standard though, it's now often the default, as its performance penalty decreased
+significantly.
+
+## The Notation
+
+Much like how humans use [scientific notation](https://en.wikipedia.org/wiki/Scientific_notation) (e.g. `6.022 * 10^23`)
+for expressing numbers of arbitrary magnitude in a standard way, computers under the hood store each floating-point
+value as three numbers cleverly put together:
+
+```
+(-1)^sign * significand * 2^exponent
+```
+
+-   The **sign** is a single bit – 0 if the number is positive, 1 if negative.
+-   The **exponent** is a signed integer. In a way, it establishes the magnitude, e.g. an exponent of 8 means that
+    the absolute value of the number must be in the range `(256,512]` (`(2^8,2^9]`).  
+    Curiously, despite the exponent being signed, it's stored not using
+    [two's complement](https://en.wikipedia.org/wiki/Two%27s_complement) but as an unsigned integer, with the computer
+    _knowing_ to subtract a known **bias** to get the real value. The raw stored value is called the **biased
+    exponent**. That bias is 127 in single-precision and 1023 in double.  
+    The exponent's size determines the _range_ of the type – 8 bits of it in single-precision and 11 in double.
+-   The **significand** (also called the *mantissa*) is a fixed-point number in the range `(1,2]` (e.g. 1 or 1.0625 or
```

Comment:
nie powinna być inna kolejność? najpierw significand a potem exponent tak jak we wzorze wyżej